backend/coins/eth/account.go

```diff
@@ -812,3 +814,133 @@ func (account *Account) VerifyAddress(addressID string) (bool, error) {
+	if err != nil {
+		return "", err
+	}
+	return "0x" + hex.EncodeToString(signedMessage), nil
+}
+
+// SignTypedMsg signs an Ethereum EIP-612 typed message in BBApp via WalletConnect
+func (account *Account) SignTypedMsg(
+	chainId uint64,
+	data string,
+) (string, error) {
+	signedMessage, err := account.Config().Keystore.SignETHTypedMessage(chainId, []byte(data), account.signingConfiguration.AbsoluteKeypath())
+	if err != nil {
+		return "", err
+	}
+	return "0x" + hex.EncodeToString(signedMessage), nil
+}
+
+// TX proposal arguments received from Wallet Connect with Gas, GasPrice, Value and Nonce being optional
+type WalletConnectArgs struct {
+	From     string `json:"from"`
+	To       string `json:"to"`
+	Data     string `json:"data"`
+	Gas      string `json:"gas,omitempty"`
+	GasPrice string `json:"gasPrice,omitempty"`
+	Value    string `json:"value,omitempty"`
+	Nonce    string `json:"nonce,omitempty"`
+}
+
+// EthSignTypedMsg signs an Ethereum Tx received from WalletConnect
+func (account *Account) EthSignWalletConnectTx(
+	// send: whether transaction should be broadcast after signing
+	send bool,
+	// chainId: allow specifying other IDs than 1 (ETH mainnet) for other EVM networks
+	chainId uint64,
+	proposedTx WalletConnectArgs,
+) (string, string, error) {
+	var nonce uint64
+	var message ethereum.CallMsg
+	var gasPrice *big.Int
+	var value *big.Int
+
+	if !ethcommon.IsHexAddress(proposedTx.To) {
+		return "", "", errp.WithStack(errors.ErrInvalidAddress)
+	}
+	address := ethcommon.HexToAddress(proposedTx.To)
+	if isMixedCase(proposedTx.To) && proposedTx.To != address.Hex() {
+		return "", "", errp.WithStack(errors.ErrInvalidAddress)
+	}
+
+	if proposedTx.Nonce != "" {
+		parsed, err := strconv.ParseUint(strings.TrimPrefix(proposedTx.Nonce, "0x"), 16, 64)
+		if err != nil {
+			return "", "", err
+		}
+		nonce = parsed
+	} else {
+		nonce = account.nextNonce
+	}
+
+	if proposedTx.Value != "" {
+		bigIntValue := new(big.Int)
+		bigIntValue.SetString(strings.TrimPrefix(proposedTx.Value, "0x"), 16)
+		value = bigIntValue
+	}
+
+	data, err := hex.DecodeString(strings.TrimPrefix(proposedTx.Data, "0x"))
+	if err != nil {
+		return "", "", err
+	}
+
+	message = ethereum.CallMsg{
+		From:     account.address.Address,
+		To:       &address,
+		Gas:      0,
+		GasPrice: big.NewInt(0),
+		Value:    value,
+		Data:     data,
+	}
+
+	gasLimit, err := account.coin.client.EstimateGas(context.TODO(), message)
+	if err != nil {
+		if strings.Contains(err.Error(), etherscan.ERC20GasErr) {
+			return "", "", errp.WithStack(errors.ErrInsufficientFunds)
+		}
+		account.log.WithError(err).Error("Could not estimate the gas limit.")
+		return "", "", errp.WithStack(errors.TxValidationError(err.Error()))
+	}
+
+	for _, t := range account.feeTargets() {
+		//TODO Let user choose gas price/priority
+		if t.code == accounts.FeeTargetCodeNormal {
```

Comment:
if this fee target does not exist you should probably error? not sure what happens if you proceed here with a 0 gas price.

Reply:
Currently this should always exist afaik
Since ethgasstation doesn't work anymore, the fallback always sets FeeTargetCodeNormal
https://github.com/digitalbitbox/bitbox-wallet-app/blob/3fb59738e0c570bb3b5ba8817479df2c2a745405/backend/coins/eth/account.go#L725

This is to be changed after adding custom gas price to WC transactions (TODO L#917)